<a href="https://colab.research.google.com/github/gongwon65/day1/blob/master/practice_numpy_1_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# (실습) 넘파이 어레이

**주의사항**

- `None`은 적절한 표현식으로 대체한다.
- `Pass`는 적절한 명령문으로 대체한다.

## numpy 모듈

**문제 1**

아래 코드를 설명하라.

In [1]:
import numpy as np

`코드설명`: numpy모듈을 별칭np로 부른다.

아래 코드는 수정하지 마세요!!!

In [2]:
np.__version__

'1.26.4'

**문제 2**

아래 코드의 기능을 설명하라

In [3]:
np.random.seed(42)

`코드설명`: 난수를 생성할 때 사용되는 시드를 42로 지정한다.

**문제 3**

부동소수점으로 구성된 어레이의 항목은 기본적으로 소수점 이하 8자리까지 출력한다.

In [4]:
Z = np.random.random((40,40))
Z

array([[0.37454012, 0.95071431, 0.73199394, ..., 0.09767211, 0.68423303,
        0.44015249],
       [0.12203823, 0.49517691, 0.03438852, ..., 0.07404465, 0.35846573,
        0.11586906],
       [0.86310343, 0.62329813, 0.33089802, ..., 0.18657006, 0.892559  ,
        0.53934224],
       ...,
       [0.81783422, 0.12020906, 0.5444891 , ..., 0.89803055, 0.2434823 ,
        0.92703454],
       [0.06026739, 0.93443603, 0.35162269, ..., 0.41102847, 0.8398613 ,
        0.90002312],
       [0.35342138, 0.23687056, 0.78052552, ..., 0.07941937, 0.08565825,
        0.89419087]])

`np.set_printoptions()` 의 기능을 다음 두 개의 예를 이용하여 설명하라.

In [5]:
np.set_printoptions(precision=4)

In [ ]:
Z

array([[0.3745, 0.9507, 0.732 , ..., 0.0977, 0.6842, 0.4402],
       [0.122 , 0.4952, 0.0344, ..., 0.074 , 0.3585, 0.1159],
       [0.8631, 0.6233, 0.3309, ..., 0.1866, 0.8926, 0.5393],
       ...,
       [0.8178, 0.1202, 0.5445, ..., 0.898 , 0.2435, 0.927 ],
       [0.0603, 0.9344, 0.3516, ..., 0.411 , 0.8399, 0.9   ],
       [0.3534, 0.2369, 0.7805, ..., 0.0794, 0.0857, 0.8942]])

In [6]:
np.set_printoptions(precision=6)

In [7]:
Z

array([[0.37454 , 0.950714, 0.731994, ..., 0.097672, 0.684233, 0.440152],
       [0.122038, 0.495177, 0.034389, ..., 0.074045, 0.358466, 0.115869],
       [0.863103, 0.623298, 0.330898, ..., 0.18657 , 0.892559, 0.539342],
       ...,
       [0.817834, 0.120209, 0.544489, ..., 0.898031, 0.243482, 0.927035],
       [0.060267, 0.934436, 0.351623, ..., 0.411028, 0.839861, 0.900023],
       [0.353421, 0.236871, 0.780526, ..., 0.079419, 0.085658, 0.894191]])

`코드설명`: np의 소수점 자릿수 출력설정을 바꾼다. 디폴트의 경우 소수점 이하 8자리, 전자의 경우 4자리, 후자의 경우 6자리가 출력된다.

아래 코드는 기본 옵션값으로 되돌린다. 수정하지 마세요.

In [8]:
np.set_printoptions(precision=8)

## 1차원 어레이

**문제 1**

크기가 10이고 모든 항목이 0인 1차원 어레이를
변수 `Z`가 가리키도록 하라.
단, `np.zeros()` 함수를 반드시 이용한다.

In [63]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros(10)

아래 코드는 수정하지 마세요!!!

In [64]:
assert (Z == np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])).all()

**문제 2**

변수 `Z`가 10에서 49까지의 정수 중에서 짝수로만 구성된 1차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수를 반드시 이용한다.

In [65]:
# None을 적절한 표현식으로 대체하라.
Z = np.arange(10,50,2)

아래 코드는 수정하지 마세요!!!

In [66]:
assert (Z == np.array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42,
                        44, 46, 48])).all()

**문제 3**

변수 `Z`가 아래와 같이 선언되었다.

In [67]:
Z = np.array([2, 3, 1, 0, 4])
Z

array([2, 3, 1, 0, 4])

변수 `Z`의 항목을 정렬하라.
단, sort() 메서드를 활용한다.

In [68]:
# None을 적절한 표현식으로 대체하라.
Z.sort()

아래 코드는 수정하지 마세요!!!

In [69]:
assert (Z == np.arange(5)).all()

**문제 4**

변수 `Z`가 크기 30이며 무작위로 생성된 1차원 어레이를 가리키도록 하라.
또한 포함된 항목들의 평균값을 계산하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
`np.random.random()` 함수, `mean()` 어레이 메서드를 반드시 이용한다.

In [70]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random(30) # 0~1 사이의 난수 생성
Z_mean = Z.mean()

아래 코드는 수정하지 마세요!!!

In [71]:
assert (Z.shape == (30,))
print(f"1차원 어레이 평균값: {Z_mean}")

1차원 어레이 평균값: 0.5274843769736939


**문제 5**

1차원 어레이가 다음과 같이 주어졌다.

In [72]:
Z = np.arange(1, 4)

(1) 아래 표현식의 계산 결과를 설명하라.

In [73]:
Z*Z

array([1, 4, 9])

`코드설명`: Z 어레이 각 객체가 제곱됨.



(2) 아래 표현식의 계산 결과를 설명하라.

In [74]:
Z**Z

array([ 1,  4, 27])

`코드설명`:객체를 지수승함.

(3) 아래 표현식의 계산 결과를 설명하라.

In [75]:
Z < -Z

array([False, False, False])

`코드설명`: 항목별 크기 비교교

(4) 아래 표현식의 계산 결과를 설명하라.

In [76]:
Z/1/1

array([1., 2., 3.])

`코드설명`: 항목별 2번 나눗셈을 진행행.

**문제 6**

변수 `Z`가 아래 모양의 어레이를 가리키도록 하라.
단, `np.linspace()` 함수와 `endpoint` 키워드 인자를 이용한다.

```
[0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9]
```

In [90]:
# None을 적절한 표현식으로 대체하라.
Z = np.linspace(0., 1., 10, endpoint=0) 
#False시 1을 포함 안함, True시 1을 포함함.

아래 코드는 수정하지 마세요!!!

In [78]:
assert np.isclose(Z, np.array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])).all()

**문제 7**

아래 코드의 실행결과를 설명하라.

In [ ]:
Z = np.zeros((10,10))
print("%d bytes" % (Z.size * Z.itemsize))

[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
800 bytes


`코드설명`: 행10 열10 총 100size, float64 각8바이트트 | 결과값 = 800 bytes

**문제 8**

다음 리스트에서 0이 아닌 항목들의 인덱스로 구성된 어레이를 한줄 코드로 생성하라.
단. `np.nonzero()` 함수를 활용한다.

In [80]:
Z = [1, 2, 0, 0, 4, 0]

In [ ]:
# None을 적절한 표현식으로 대체하라.

non_zeros = np.nonzero(Z) #nonzero인 인덱스 반환

(array([], dtype=int64), array([], dtype=int64))


아래 코드는 수정하지 마세요!!!

In [82]:
assert (non_zeros == np.array([0, 1, 4])).all()

**문제 9**

두 개의 1차원 어레이가 다음과 같이 주어졌다.

In [83]:
A = np.ones(3)*1
B = np.ones(3)*2

아래 코드의 실행 결과를 설명하라.

In [84]:
(A+B)*(-A/2)

array([-1.5, -1.5, -1.5])

`코드설명`: (1+2)*(-0.5) = -1.5

In [85]:
np.add(A,B,out=B)

array([3., 3., 3.])

`코드설명`: A와 B를 더하고 그 값을 B에다 덮어씌움

In [86]:
np.divide(A,2,out=A)

array([0.5, 0.5, 0.5])

`코드설명`: A를 2로 나누고 그 값을 A에 저장

In [87]:
np.negative(A,out=A)

array([-0.5, -0.5, -0.5])

`코드설명`: A의 부호를 반전시키고 그 값을 A에 저장

In [88]:
np.multiply(A,B,out=A)

array([-1.5, -1.5, -1.5])

`코드설명`: A와 B를 곱한 뒤 A에 저장

## 2차원 어레이

**문제 1**

변수 `Z`가 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random((5,5))

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Z.shape == (5, 5))
assert (Z.dtype == np.float64)

**문제 2**

변수 `Z`가 0에서 8까지의 정수로 이루어진 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `reshape()` 어레이 메서드를 반드시 이용한다.

In [99]:
# None을 적절한 표현식으로 대체하라.
Z = np.arange(9).reshape(3,3)

아래 코드는 수정하지 마세요!!!

In [100]:
assert (Z == np.array([[0, 1, 2],
                       [3, 4, 5],
                       [6, 7, 8]])).all()

**문제 3**

변수 `Z`가 (3, 3) 모양의 2차원 어레이를 가리키도록 하라.
단, 2차원 어레이의 대각선 상은 1이고 나머지는 0이어야 한다.
또한 `np.eye()` 함수를 반드시 이용한다.

In [101]:
# None을 적절한 표현식으로 대체하라.
Z = np.eye(3)

아래 코드는 수정하지 마세요!!!

In [102]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all

**문제 4**

변수 `Z`가 아래와 같은 (5,5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.arange()` 함수와 `np.diag()` 함수의 `k` 키워드 인자를 반드시 활용한다.

```
[[0 0 0 0 0]
 [1 0 0 0 0]
 [0 2 0 0 0]
 [0 0 3 0 0]
 [0 0 0 4 0]]
```

In [ ]:
# None을 적절한 표현식으로 대체하라.
Z = np.diag(1+np.arange(4), k=-1) #1부터 시작 k는 대각선의 위치. 2차원 어레이 생성성

아래 코드는 수정하지 마세요!!!

In [106]:
assert (Z == np.array([[0, 0, 0, 0, 0],
                        [1, 0, 0, 0, 0],
                        [0, 2, 0, 0, 0],
                        [0, 0, 3, 0, 0],
                        [0, 0, 0, 4, 0]])).all()

**문제 5**

변수 `Z`가 아래와 같은 (5, 5) 모양의 2차원 어레이를 가리키도록 하라.
단, `np.zeros()` 함수, `np.arange()` 함수, 2차원 어레이 덧셈을 이용한다.

```
[[0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]
 [0. 1. 2. 3. 4.]]
```

In [110]:
# None을 적절한 표현식으로 대체하라.
Z = np.zeros((5,5)) + np.arange(5)

아래 코드는 수정하지 마세요!!!

In [111]:
assert (Z == np.array([[0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.],
                        [0., 1., 2., 3., 4.]])).all()

**문제 6**

아래 3x3 단위 행렬에 해당하는 2차원 어레이를 한줄 코드로 생성하라.
단, `np.eye()` 함수를 이용한다.

```
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
```

In [115]:
# None을 적절한 표현식으로 대체하라.
Z = np.eye(3)
print(Z)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


아래 코드는 수정하지 마세요!!!

In [116]:
assert (Z == np.array([[1., 0., 0.],
                       [0., 1., 0.],
                       [0., 0., 1.]])).all()

**문제 7**

무작위 수로 구성된 (5, 5) 모양의 2차원 어레이가 다음과 같다.

In [ ]:
#지정한 범위로 값을 조정하는 것: 정규화, -1~1 또는 0~1 대부분이 후자 
#대원칙: 인공지능은 학습전에 데이터가 정규화 되어있어야 한다. (안되어있어도 되지만 성능이 딸림)
Z = np.random.random((5, 5))

변수 `Z`가 가리키는 2차원 어레이에 대해 min-max 크기조정(정규화)을 적용하라.

min-max 크기조정(정규화)은 항목의 모든 값을 아래 식을 통해 0과 1 사이의 부동소수점으로 변환한다. $max$와 $min$은 각각 최대 항목과 최소 항목을 가리킨다.

$$
\frac{x-min}{max - min}
$$

In [117]:
# None을 적절한 표현식으로 대체하라.
Z = (Z-Z.min())/(Z.max()-Z.min())

아래 코드는 수정하지 마세요!!!

In [118]:
assert (Z.max() == 1 and Z.min() == 0)

## 3차원 어레이

**문제 1**

(1) 변수 `Z`가 (3, 3, 3) 모양의 3차원 어레이를 가리키도록 하라.
단, 어레이의 항목은 0과 1 사이의 임의의 부동소수점으로 구성되어야 한다.
또한 `np.random.random()` 함수를 이용한다.

In [121]:
# None을 적절한 표현식으로 대체하라.
Z = np.random.random((3,3,3))
print(Z)

[[[0.93738807 0.04317374 0.16481482]
  [0.13172877 0.72597991 0.81778533]
  [0.21351137 0.50585269 0.84070303]]

 [[0.73280154 0.54223721 0.59034769]
  [0.50836055 0.29754845 0.56502199]
  [0.6888853  0.87332292 0.63629135]]

 [[0.76112154 0.16007163 0.46155747]
  [0.00933162 0.24667886 0.72646172]
  [0.99180995 0.0991781  0.40149432]]]


아래 코드는 수정하지 마세요!!!

In [122]:
assert (Z.shape == (3, 3, 3))
assert (Z.dtype == 'f8')

(2) `Z`에 포함된 항목들 중에서 최댓값과 최솟값의 차이를 구하라.
단, `maximum`, `mininum` 두 변수가 각각 최댓값과 최솟값을 가리키도록 하며,
`max()`와 `min()` 어레이 메서드를 이용한다.

In [123]:
# None을 적절한 표현식으로 대체하라.
maximum = Z.max()
minimum = Z.min()

아래 코드는 수정하지 마세요!!!

In [124]:
print(f"최댓값과 최솟값의 차이: {maximum - minimum}")

최댓값과 최솟값의 차이: 0.9824783315222702


## `dtype` 종류

**문제 1**

아래 두 코드의 실행결과를 설명하라.

In [125]:
for dtype in [np.int8, np.int32, np.int64]:
    print(np.iinfo(dtype).min)
    print(np.iinfo(dtype).max)

-128
127
-2147483648
2147483647
-9223372036854775808
9223372036854775807


`코드설명`: 정수형 자료형에 허용되는 비트수에 따른 최소값과 최대값. #2^8=256 이기에 0포함 255. -128~127까지

In [126]:
for dtype in [np.float32, np.float64]:
    print(np.finfo(dtype).min)
    print(np.finfo(dtype).max)
    print(np.finfo(dtype).eps)

-3.4028235e+38
3.4028235e+38
1.1920929e-07
-1.7976931348623157e+308
1.7976931348623157e+308
2.220446049250313e-16


`코드설명`: 부동소수점 자료형에 허용되는 비트수에 따른 최소값과 최대값값

**문제 2**

부동소수점으로 구성된 1차원 어레이가 아래와 같이 주어졌다.

In [129]:
Z = (np.random.rand(10)*100).astype(np.float32)
print(Z)

[99.63343    36.83095    44.861065   72.20709    88.619576   59.304432
 39.15257    41.262184   69.56181     0.32182637]


변수 `Y`가 변수 `Z`가 가리키는 1차원 어레이의 dtype이 `int32`로 형변환된 1차원 어레이를 가리키도록 하라.
단, `astype()` 메서드를 이용한다.

In [131]:
# None을 적절한 표현식으로 대체하라.
Y = Z.astype(np.int32)
print(Y)

[99 36 44 72 88 59 39 41 69  0]


아래 코드는 수정하지 마세요!!!

In [132]:
assert (Y.shape == (10,))
assert (Y.dtype == np.int32)

**문제 3**

변수 `Z`가 아래와 같이 무작위로 생성된 길이가 10인 1차원 어레이를 가리킨다.

In [133]:
Z = np.random.uniform(0,10,10) #0부터10사이 숫자를 10개 생성
Z

array([6.19589337, 3.5549301 , 7.94197331, 0.9299064 , 5.88202267,
       4.8097289 , 6.42325526, 0.64853594, 5.79983787, 5.61484588])

(1) 나머지 연산자(`%`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [134]:
# None을 적절한 표현식으로 대체하라.
Y = Z - Z%1

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Y == Z.astype(np.int32)).all()

(2) 몫 연산자(`//`)를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [135]:
# None을 적절한 표현식으로 대체하라.
Y = Z//1

아래 코드는 수정하지 마세요!!!

In [ ]:
assert (Y == Z.astype(np.int32)).all()

(3) 내림 함수 `np.floor()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [136]:
# None을 적절한 표현식으로 대체하라.
Y = np.floor(Z)
print(Y)

[6. 3. 7. 0. 5. 4. 6. 0. 5. 5.]


아래 코드는 수정하지 마세요!!!

In [137]:
assert (Y == Z.astype(np.int32)).all()

(4) `np.trunc()` 함수를 이용하여 정수로만 구성된 어레이 `Y`를 생성하라.

In [139]:
# None을 적절한 표현식으로 대체하라.
Y = np.trunc(Z)
print(Y)

[6. 3. 7. 0. 5. 4. 6. 0. 5. 5.]


아래 코드는 수정하지 마세요!!!

In [140]:
assert (Y == Z.astype(np.int32)).all()